### PR TITLE
Dynamic memory audit

### PR DESCRIPTION
**Pull request by wjwwood**
This pull request makes sure that we're using `rcutils_allocator_t` everywhere and not using `malloc`, `realloc`, `calloc`, or `free` anywhere at all.

Requires #101

### CODE REVIEW

**Comment by dejanpan** [comment]
@wjwwood sorry but I can't recall it now - you said that there will be another PR against rcutils after this one?

**Comment by wjwwood** [comment]
Yes, I will probably make it another one rather than combining it into this.

**Comment by wjwwood** [comment]
If anyone has time to review these, that would be appreciated! 👍

CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/4687/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/1574/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/3885/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/4763/)

**Comment by dejanpan** [comment]
@serge-nikulin could you do it please?

File: include/rcutils/concat.h [hunk 1]
```diff
@@ -1,42 +0,0 @@
-// Copyright 2017 Open Source Robotics Foundation, Inc.
```
Comment (dejanpan) [inline include/rcutils/concat.h]: why was this file removed?

I tried to find any mention but I couldn't.

File: test/test_logging_macros.c [hunk 2]
```diff
@@ -32,15 +33,16 @@ void custom_handler(
...
   g_last_log_event.location = location;
   g_last_log_event.severity = severity;
   g_last_log_event.name = name ? name : "";
   if (g_last_log_event.message) {
-    free(g_last_log_event.message);
+    allocator.deallocate(g_last_log_event.message, allocator.state);
   }
   const size_t size = 1024;
-  g_last_log_event.message = malloc(size);
+  g_last_log_event.message = allocator.allocate(size, allocator.state);
```
Comment (dejanpan) [inline test/test_logging_macros.c]: so `free` resp. `malloc` are now called inside `deallocate` resp `allocate`?

File: include/rcutils/concat.h (hunk 1 above)
Comment (wjwwood) [reply include/rcutils/concat.h]: I just replaced all use of it with `rcutils_format_string()`, so it's unnecessary now.

File: test/test_logging_macros.c (hunk 2 above)
Comment (wjwwood) [reply test/test_logging_macros.c]: Yes.

**Comment by wjwwood** [comment]
Sorry, there are some missing, but related pull requests for this one. I'll fix it up and ask for review again.

**Comment by dejanpan** [comment]
@wjwwood one more Q. I tried to find coverage information for rcutils but could only find this https://ci.ros2.org/job/ci_linux_coverage/2/cobertura/src_ros2_rcutils_src/ which seems a bit old.

**Comment by wjwwood** [comment]
@serge-nikulin you're welcome to review it too, but I do need someone from our team to review it as well.

> @wjwwood one more Q. I tried to find coverage information for rcutils but could only find this https://ci.ros2.org/job/ci_linux_coverage/2/cobertura/src_ros2_rcutils_src/ which seems a bit old.

There's no coverage information still. We have what you linked to, but it's also not trusty worthy right now.

**Comment by serge-nikulin** [comment]
@dejanpan, when do you need the review?

File: test/test_filesystem.cpp [hunk 3]
```diff
@@ -19,8 +19,10 @@
 
 static char cwd[1024];
 
+static rcutils_allocator_t g_allocator = rcutils_get_default_allocator();
```
[the remark this answers is not in the document]
Comment (wjwwood) [reply test/test_filesystem.cpp]: Yeah, looks like there are some memory leaks in that test, I'll clean them up.

[the remark this answers is not in the document]
Comment (wjwwood) [reply test/test_filesystem.cpp]: Yeah, those were really bad leaks. I think 717e044 fixed them.

**Comment by wjwwood** [comment]
Ok, in review again, new CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/4691/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/1578/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/3892/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/4769/)

**Comment by wjwwood** [comment]
I could still use reviews of some of the other pr's associated with this pull request.

New CI after fix up:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/4704/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/1584/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/3903/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/4776/)
